From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Llewellyn-Jones <david@flypig.co.uk>
Date: Wed, 23 Nov 2022 09:08:22 +0200
Subject: [PATCH 49/67] Add support for prefers-color-scheme JB#58394

Adds support to qt/nsLookAndFeel for the prefers-color-scheme media tag.
The renderer already supports it by setting the ui.systemUsesDarkTheme
preference to 0 or 1.

This patch allows, in addition, for the colour scheme to be controlled
by sending a "ambience-theme-changed" notification with a data payload
of either "light" or "dark". If the ui.systemUsesDarkTheme preference is
unset or set to something other than 0 or 1, the notification value will
be used to control the theme instead.

This functionality is used to update the colour scheme based on the
ambience of the device.
---
 widget/nsXPLookAndFeel.cpp | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/widget/nsXPLookAndFeel.cpp b/widget/nsXPLookAndFeel.cpp
index fae1444161b4..15132d713030 100644
--- a/widget/nsXPLookAndFeel.cpp
+++ b/widget/nsXPLookAndFeel.cpp
@@ -1042,8 +1042,10 @@ nsresult nsXPLookAndFeel::GetIntValue(IntID aID, int32_t& aResult) {
   }
 
   if (NS_SUCCEEDED(Preferences::GetInt(sIntPrefs[size_t(aID)], &aResult))) {
-    sIntCache.Insert(aID, Some(aResult));
-    return NS_OK;
+    if ((aID != IntID::SystemUsesDarkTheme) || (aResult != 2)) {
+      sIntCache.Insert(aID, Some(aResult));
+      return NS_OK;
+    }
   }
 
   if (NS_FAILED(NativeGetInt(aID, aResult))) {
